### PR TITLE
Add a new wasm32-unknown-wasi target

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds a new wasm32-based target distributed through rustup,
supported in the standard library, and implemented in the compiler. The
`wasm32-unknown-wasi` target is intended to be a WebAssembly target
which matches the [WASI proposal recently announced][LINK]. In summary
the WASI target is an effort to define a standard set of syscalls for
WebAssembly modules, allowing WebAssembly modules to not only be
portable across architectures but also be portable across environments
implementing this standard set of system calls.

The wasi target in libstd is still somewhat bare bones. This PR does not
fill out the filesystem, networking, threads, etc. Instead it only
provides the most basic of integration with the wasi syscalls, enabling
features like:

* `Instant::now` and `SystemTime::now` work
* `env::args` is hooked up
* `env::vars` will look up environment variables
* `println!` will print to standard out
* `process::{exit, abort}` should be hooked up appropriately

None of these APIs can work natively on the `wasm32-unknown-unknown`
target, but with the assumption of the WASI set of syscalls we're able
to provide implementations of these syscalls that engines can implement.
Currently the primary engine implementing wasi is [wasmtime], but more
will surely emerge!

In terms of future development of libstd, I think this is something
we'll probably want to discuss. The purpose of the WASI target is to
provide a standardized set of syscalls, but it's *also* to provide a
standard C sysroot for compiling C/C++ programs. This means it's
intended that functions like `read` and `write` are implemented for this
target with a relatively standard definition and implementation. It's
unclear, therefore, how we want to expose file descriptors and how we'll
want to implement system primitives. For example should `std::fs::File`
have a libc-based file descriptor underneath it? The raw wasi file
descriptor? We'll see! Currently these details are all intentionally
hidden and things we can change over time.

A `WasiFd` sample struct was added to the standard library as part of
this commit, but it's not currently used. It shows how all the wasi
syscalls could be ergonomically bound in Rust, and they offer a possible
implementation of primitives like `std::fs::File` if we bind wasi file
descriptors exactly.

Apart from the standard library, there's also the matter of how this
target is integrated with respect to its C standard library. The
reference sysroot, for example, provides managment of standard unix file
descriptors and also standard APIs like `open` (as opposed to the
relative `openat` inspiration for the wasi ssycalls). Currently the
standard library relies on the C sysroot symbols for operations such as
environment management, process exit, and `read`/`write` of stdio fds.
We want these operations in Rust to be interoperable with C if they're
used in the same process. Put another way, if Rust and C are linked into
the same WebAssembly binary they should work together, but that requires
that the same C standard library is used.

We also, however, want the `wasm32-unknown-wasi` target to be
usable-by-default with the Rust compiler without requiring a separate
toolchain to get downloaded and configured. With that in mind, there's
two modes of operation for the `wasm32-unknown-wasi` target:

1. By default the C standard library is statically provided inside of
   `liblibc.rlib` distributed as part of the sysroot. This means that
   you can `rustc foo.wasm --target wasm32-unknown-unknown` and you're
   good to go, a fully workable wasi binary pops out. This is
   incompatible with linking in C code, however, which may be compiled
   against a different sysroot than the Rust code was previously
   compiled against. In this mode the default of `rust-lld` is used to
   link binaries.

2. For linking with C code, the `-C target-feature=-crt-static` flag
   needs to be passed. This takes inspiration from the musl target for
   this flag, but the idea is that you're no longer using the provided
   static C runtime, but rather one will be provided externally. This
   flag is intended to also get coupled with an external `clang`
   compiler configured with its own sysroot. Therefore you'll typically
   use this flag with `-C linker=/path/to/clang-script-wrapper`. Using
   this mode the Rust code will continue to reference standard C
   symbols, but the definition will be pulled in by the linker configured.

Alright so that's all the current state of this PR. I suspect we'll
definitely want to discuss this before landing of course! This PR is
coupled with libc changes as well which I'll be posting shortly.

[LINK]: https://hacks.mozilla.org/2019/03/standardizing-wasi-a-webassembly-system-interface/
[wasmtime]: https://github.com/cranestation/wasmtime-wasi

### CODE REVIEW

**Comment by alexcrichton** [comment]
This is coupled with https://github.com/rust-lang/libc/pull/1307 for the `libc` crate changes, and we'll need to publish that before merging this.

r? @fitzgen

File: src/bootstrap/compile.rs
Comment (fitzgen) [inline src/bootstrap/compile.rs]:
```suggestion
    } else if target.ends_with("-wasi") {
```

?

File: src/bootstrap/compile.rs
Comment (fitzgen) [inline src/bootstrap/compile.rs]:
```suggestion
        if target.ends_with("-wasi") {
```

File: src/librustc_target/spec/wasm32_unknown_wasi.rs
Comment (fitzgen) [inline src/librustc_target/spec/wasm32_unknown_wasi.rs]: ☝️

```suggestion
https://github.com/CraneStation/wasmtime-wasi/blob/wasi/docs/WASI-proposed-CG-subgroup-charter.md
```

or maybe just wasi.dev?

File: src/librustc_target/spec/wasm32_unknown_wasi.rs
Comment (fitzgen) [inline src/librustc_target/spec/wasm32_unknown_wasi.rs]: Thanks for this very nice doc comment :+1:

File: src/libstd/sys/wasi/os.rs
Comment (fitzgen) [inline src/libstd/sys/wasi/os.rs]: Might make sense to have an `inline(never)` function that all the unsupported panics defer to. I think this should make code sizes smaller.

**Review by fitzgen — APPROVED** [review]
:tada:

File: src/libstd/sys/wasi/os.rs
Comment (gnzlbg) [inline src/libstd/sys/wasi/os.rs]: Since glibc is GPL'ed, it might be wiser to take inspiration from musl here.

**Comment by gnzlbg** [comment]
> For example should std::fs::File
have a libc-based file descriptor underneath it? The raw wasi file
descriptor? We'll see! Currently these details are all intentionally
hidden and things we can change over time.
>
> A WasiFd sample struct was added to the standard library as part of
this commit, but it's not currently used. It shows how all the wasi
syscalls could be ergonomically bound in Rust, and they offer a possible
implementation of primitives like std::fs::File if we bind wasi file
descriptors exactly.

I'm not sure I understand. Is there something that prevents `std::fs::File` from being implemented on top of the libc wasi file descriptor? Are you arguing that maybe we should expose a `WasiFd` instead of `std::fs::File` for this target ?

File: src/libstd/sys/wasi/os.rs
Comment (alexcrichton) [reply src/libstd/sys/wasi/os.rs]: Perhaps yeah! I figure we can fix that up as it goes along because unconditionally panicking binaries aren't too useful anyway :)

**Comment by alexcrichton** [comment]
> Is there something that prevents std::fs::File from being implemented on top of the libc wasi file descriptor? Are you arguing that maybe we should expose a WasiFd instead of std::fs::File for this target ?

This sort of depends on how the file descriptor situation is settled in wasi's sysroot. It seems like there's the concept of a "wasi file descriptor" as well as a "libc file descriptor", indexing different spaces. For those purposes we'd need to convert between them.

This is sort of similar as well to "what if we used `FILE` as the implementation of `std::fs::File` on Unix?". I think the answer for wasi is "all primitives should internally contain `WasiFd`", but we just need to figure out a way how to do that which still interoperates with C code well.

File: src/ci/docker/dist-various-2/build-wasi-toolchain.sh
Comment (chicoxyzzy) [inline src/ci/docker/dist-various-2/build-wasi-toolchain.sh]: Might make sense to use latest LTS (18.04) instead

**Comment by sunfishcode** [comment]
> This sort of depends on how the file descriptor situation is settled in wasi's sysroot. It seems like there's the concept of a "wasi file descriptor" as well as a "libc file descriptor", indexing different spaces. For those purposes we'd need to convert between them.

WASI uses `__wasi_fd_t` which is unsigned, while libc/POSIX use `int` which is signed so that it can use negative numbers to indicate errors, but it's the same index space.

File: src/ci/docker/dist-various-2/build-wasi-toolchain.sh
Comment (alexcrichton) [reply src/ci/docker/dist-various-2/build-wasi-toolchain.sh]: We will eventually! For now though the rest of the image here is 17.10, but when that upgrades we can upgrade this

**Comment by alexcrichton** [comment]
@sunfishcode oh ok thanks! So the idea is that a libc file descriptor is interchangeable with a wasi file descriptor? If that's the case sounds like we'll be using a design similar to unix for libstd where every type just internally contains a 32-bit file descriptor (probably the `WasiFd` type in this PR)

**Comment by sunfishcode** [comment]
Yes, they're interchangeable. (If WASI ever produces an fd greater than 2<sup>31</sup>, libc might get confused, but realistically you'll hit other limits well before that happens.)

**Comment by Centril** [comment]
For rollup planning, @bors p=2

**Pull request merged**
